docs/nidigital/class.rst

```diff
@@ -1557,6 +1557,66 @@ export_signal
 
             :type output_terminal: str
 
+fetch_capture_waveform
+----------------------
+
+    .. py:currentmodule:: nidigital.Session
+
+    .. py:method:: fetch_capture_waveform(site_list, waveform_name, samples_to_read, timeout)
+
+            Returns a list of named tuples (Waveform) that <FILL IN THE BLANK HERE>
+
+            Fields in Waveform:
+
+            - **site** (int)
+            - **data** (array.array of int)
```

Comment:
I wonder if named tuple of int/array of int is the best way to return this.
What are all the places in the API in which this type could be used? Can it ever be an input?

Reply:
@sbethur?

Reply:
It's used in two places:

1) `niDigital_SortSiteResultsViUInt32Waveform` - This will NOT be added in Python, since its use is superseded by multi-instrument sessions.
2) `niDigital_WriteSourceWaveformSiteUniqueU32` - This should use the same type as what we use here for `fetch_capture_waveform`.  Tracked [here](https://github.com/ni/nimi-python/issues/1069).

I looked at one of the use-cases for `fetch_capture_waveform` - streaming capture waveform data from h/w test program. It would be ideal for `fetch_capture_waveform` to accept a buffer and populate that with the waveform data (like [fetch_into](https://nimi-python.readthedocs.io/en/master/niscope/class.html#fetch-into) in niscope). 

But the streaming use-case is not common and even then the size is about tens of MBs. If we learn about a use-case where performance is not sufficient, we can add `fetch_into` later. Usability of streaming with `fetch_capture_waveform` that returns a dict is not bad:
```
# Pattern burst is configured to fetch 1024 samples
 num_samples_expected = 16
 samples_per_fetch = 4
 waveforms = collections.defaultdict(list)
 for i in range(num_samples_expected // samples_per_fetch):
     fetched_waveform = multi_instrument_session.fetch_capture_waveform(site_list='', waveform_name='capt_wfm', samples_to_read=samples_per_fetch, timeout=10.0)
     for key in fetched_waveform:
         waveforms[key] += fetched_waveform[key]
```

So, now the question is whether to use dict or namedtuple..

One drawback for dict is iteration order is not deterministic (at least in Python 3.5). It looks like in 3.6 and later, iteration is ordered. Either way, I think it's better to not guarantee any order, because trying to do that is little tricky:
 
If method is called with `site=''` then the order will 0,1,2... with disabled sites skipped.
If method is called with `site='3,0,1'` (crazy but allowed), then the order will be 3,0,1 with disabled sites skipped.

Given that order depends on input and also can contain gaps, users should not really rely on order. Even if they want to iterate in a particular order, it's not difficult to do that in Python (Everything's easy in Python )

In summary, the recommendation is dict(int, array of unsigned int)